map/separate_chaining_hashtable.py

```diff
@@ -0,0 +1,137 @@
+import unittest
+from unittest import TestCase
+
+
+class Node(object):
+    key = None
+    value = None
+    next = None
+
+    def __init__(self, key=None, value=None, next=None):
+        self.key = key
+        self.value = value
+        self.next = next
+
+
+class SeparateChainingHashTable(object):
+    """
+    By having each bucket contain a linked list of elements that are hashed to that bucket.
+    """
+    _empty = None
+
+    def __init__(self, size=11):
+        self.size = size
+        self._len = 0
+        self._table = [self._empty] * size
+
+    def put(self, key, value):
+        hash_ = self.hash(key)
+        node_ = self._table[hash_]
+        if node_ is self._empty:
+            self._table[hash_] = Node(key, value)
+            self._len += 1
+            return
+        while node_.next is not None:
+            if node_.key == key:
+                node_.value = value
+                return
+            node_ = node_.next
+        node_.next = Node(key, value)
+        self._len += 1
+
+    def get(self, key):
+        hash_ = self.hash(key)
+        node_ = self._table[hash_]
+        while node_ is not self._empty:
+            if node_.key == key:
+                return node_.value
+            node_ = node_.next
+        return None
+
+    def del_(self, key):
+        hash_ = self.hash(key)
+        node_ = self._table[hash_]
+        pre_node = None
+        while node_ is not None:
+            if node_.key == key:
+                if pre_node is None:
+                    self._table[hash_] = node_.next
+                else:
+                    pre_node.next = node_.next
+                self._len -= 1
+            pre_node = node_
+            node_ = node_.next
+
+    def hash(self, key):
+        return hash(key) % self.size
```

Comment:
Why you wrote this method recursively? Simple ```return key % self.size``` does the job, too.

Reply:
I think its not recursive. `hash()` is the builtin function here inside the defined method. Recursive would be `self.hash()`.

Reply:
@goswami-rahul Thanks for the note.